rust_icu_ubrk/src/lib.rs

```diff
@@ -0,0 +1,735 @@
+
+        assert_eq!(iter.next(), Some(51));
+        assert_eq!(iter.get_rule_status(), 99);
+        assert_eq!(iter.get_rule_status_vec().unwrap(), vec![99]);
+
+        assert_eq!(iter.next(), Some(66));
+        assert_eq!(iter.get_rule_status(), 0);
+        assert_eq!(iter.get_rule_status_vec().unwrap(), vec![0]);
+
+        assert_eq!(iter.next(), None);
+    }
+
+    #[test]
+    fn test_clone() {
+        let mut original =
+            UBreakIterator::try_new(UBRK_WORD, "en", TEXT).unwrap();
+        original.first();
+
+        assert_eq!(Rc::strong_count(&original.text), 1);
+        assert_eq!(Rc::strong_count(original.locale.as_ref().unwrap()), 1);
+
+        assert_eq!(original.next(), Some(1));
+        assert_eq!(original.next(), Some(3));
+        assert_eq!(original.current(), 3);
+
+        // Clone in a new scope.
+        {
+            let mut clone = original.safe_clone().unwrap();
+
+            assert_eq!(Rc::strong_count(&original.text), 2);
+            assert_eq!(Rc::strong_count(original.locale.as_ref().unwrap()), 2);
+
+            assert_eq!(clone.current(), 3);
+            assert_eq!(clone.first(), 0);
+            assert_eq!(clone.next(), Some(1));
+
+            assert_eq!(original.next(), Some(4));
+        }
+
+        assert_eq!(Rc::strong_count(&original.text), 1);
+        assert_eq!(Rc::strong_count(original.locale.as_ref().unwrap()), 1);
+
+        assert_eq!(original.current(), 4);
+        assert_eq!(original.next(), Some(10));
+    }
+
+    #[test]
+    fn test_set_text() {
+        let mut iter = UBreakIterator::try_new(UBRK_WORD, "en", TEXT).unwrap();
+        let original_iter_text = iter.text.clone();
+
+        assert_eq!(Rc::strong_count(&original_iter_text), 2);
+        assert_eq!(iter.preceding(59), 58);
+        assert_eq!(iter.current(), 58);
+
+        iter.set_text("The lazy dog.").unwrap();
+
+        assert_eq!(Rc::strong_count(&original_iter_text), 1);
+        assert_eq!(String::try_from(&*iter.text).unwrap(), "The lazy dog.");
+        assert_eq!(iter.current(), 0);
+        assert_eq!(iter.last_boundary(), 13);
+    }
+
+    #[test]
+    fn test_get_locale_by_type() {
+        let iter =
+            UBreakIterator::try_new(UBRK_WORD, "en_US_CA@lb=strict", TEXT)
+                .unwrap();
+
+        // The "valid locale" is the most specific locale supported by ICU, given
+        // what was requested.
+        assert_eq!(
+            iter.get_locale_by_type(ULOC_VALID_LOCALE).unwrap(),
+            "en_US"
+        );
+
+        // The "actual locale" is the locale that breaking information actually comes from.
+        // In most cases this will be "root".
+        assert_eq!(
+            iter.get_locale_by_type(ULOC_ACTUAL_LOCALE).unwrap(),
+            "root"
+        );
+    }
+
+    #[test]
+    fn test_available_locales() {
+        trace!("Available locales");
+        let count = UBreakIterator::count_available_locales();
+        for i in 0..count {
```

Comment:
what happens if `i > count`?

Reply:
This is moot following the change to replace these methods with an iterator, but I was curious. It turns out the underlying API will happily return a NULL pointer, and thus `get_available_locale_at` will return None. Safe, but inelegant.